basic_module/distribution/src/lib.rs

```diff
@@ -13,3 +13,305 @@
+        if quantity == 0 {
+            return
+        }
+        *self.current_reward.entry(*address).or_insert(0) = quantity;
+    }
+
+    pub fn drain_calculated_rewards(&mut self) -> BTreeMap<Public, u64> {
+        let mut drained = BTreeMap::new();
+        mem::swap(&mut drained, &mut self.calculated_reward);
+
+        self.save_to_state();
+        drained
+    }
+
+    pub fn update_current_rewards_to_empty(&mut self) -> Result<(), Error> {
+        let drained = BTreeMap::new();
+        self.current_reward = drained;
+
+        self.save_to_state();
+        Ok(())
+    }
+
+    pub fn get_current_rewards(&self) -> BTreeMap<Public, u64> {
+        self.current_reward.clone()
+    }
+}
+
+impl Default for IntermediateRewards {
+    fn default() -> Self {
+        IntermediateRewards {
+            current_reward: BTreeMap::default(),
+            calculated_reward: BTreeMap::default(),
+            total_additional_fee: Default::default(),
+            total_min_fee: Default::default(),
+        }
+    }
+}
+
+impl From<Vec<u8>> for IntermediateRewards {
+    fn from(_vec: Vec<u8>) -> Self {
+        IntermediateRewards::default()
+    }
+}
+
+impl From<IntermediateRewards> for Vec<u8> {
+    fn from(_reward: IntermediateRewards) -> Self {
+        Vec::default()
+    }
+}
+
+pub struct FeeDistributeIter<'a> {
+    total_stakes: u64,
+    total_min_fee: u64,
+    remaining_fee: u64,
+    stake_holdings: hash_map::Iter<'a, Public, u64>,
+}
+
+impl<'a> FeeDistributeIter<'a> {
+    pub fn remaining_fee(&self) -> u64 {
+        self.remaining_fee
+    }
+}
+
+pub fn fee_distribute<S: ::std::hash::BuildHasher>(
+    total_min_fee: u64,
+    stakes: &HashMap<Public, u64, S>,
+) -> FeeDistributeIter<'_> {
+    FeeDistributeIter {
+        total_stakes: stakes.values().sum(),
+        total_min_fee,
+        remaining_fee: total_min_fee,
+        stake_holdings: stakes.iter(),
+    }
+}
+
+impl<'a> Iterator for FeeDistributeIter<'a> {
+    type Item = (&'a Public, u64);
+    fn next(&mut self) -> Option<(&'a Public, u64)> {
+        if let Some((stakeholder, stake)) = self.stake_holdings.next() {
+            let share = share(self.total_stakes, *stake, self.total_min_fee);
+            self.remaining_fee = self.remaining_fee.checked_sub(share).expect("Remaining fee shouldn't be depleted");
+            Some((stakeholder, share))
+        } else {
+            None
+        }
+    }
+}
+
+fn share(total_stakes: u64, stake: u64, total_min_fee: u64) -> u64 {
+    assert!(total_stakes >= stake);
+    u64::try_from((u128::from(total_min_fee) * u128::from(stake)) / u128::from(total_stakes)).unwrap()
+}
+
+pub fn get_context() -> &'static mut dyn Context {
+    // This function should be implemented after the context has been formatted.
+    unimplemented!();
+}
+
+pub fn get_storage_id() -> StorageId {
+    // This function should be implemented after the storageId has been formatted.
+    unimplemented!()
+}
+
+pub fn get_key() -> H256 {
```

Comment:
You can name the key arbitrarily. Module level keys don't have to have `H256` types.